modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,327 @@
+// will be called (with all of those values as its arguments) and the output
+// will be wrapped in another Maybe.
+export const lift2 = <A, B, Out>(a: Maybe<A>, b: Maybe<B>, fn: (a: A, b: B) => Out): Maybe<Out> => {
+  if (isJust(a) && isJust(b)) {
+    return just(fn(a.value, b.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 3 maybe objects.
+export const lift3 = <A, B, C, Out>(a: Maybe<A>, b: Maybe<B>, c: Maybe<C>, fn: (a: A, b: B, c: C) => Out): Maybe<Out> => {
+  if (isJust(a) && isJust(b) && isJust(c)) {
+    return just(fn(a.value, b.value, c.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 4 maybe objects.
+export const lift4 = <A, B, C, D, Out>(
+  a: Maybe<A>,
+  b: Maybe<B>,
+  c: Maybe<C>,
+  d: Maybe<D>,
+  fn: (a: A, b: B, c: C, d: D) => Out
+): Maybe<Out> => {
+  if (isJust(a) && isJust(b) && isJust(c) && isJust(d)) {
+    return just(fn(a.value, b.value, c.value, d.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 5 maybe objects.
+export const lift5 = <A, B, C, D, E, Out>(
+  a: Maybe<A>,
+  b: Maybe<B>,
+  c: Maybe<C>,
+  d: Maybe<D>,
+  e: Maybe<E>,
+  fn: (a: A, b: B, c: C, d: D, e: E) => Out
+): Maybe<Out> => {
+  if (isJust(a) && isJust(b) && isJust(c) && isJust(d) && isJust(e)) {
+    return just(fn(a.value, b.value, c.value, d.value, e.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// --- Monad ---
+
+// Un-nests a Maybe type. Conversion is as follows:
+// Nothing ==> Nothing
+// Just(Nothing) ==> Nothing
+// Just(Just(T)) ==> Just(T)
+//
+// You shouldn't often end up with nested Maybe types to need this function, but
+// if you do it's here.
+export const flatten = <T>(self: Maybe<Maybe<T>>): Maybe<T> =>
+  isJust(self) ? self.value : nothing<T>();
+
+// Perform a transform on a Maybe object, **if** there is a value. Unlike the
+// map function earlier in the piece, here the transform itself also returns a
+// Maybe. Think of this like a combination of `map` and `flatten` at the same
+// time.
+export const bind = <T, U>(binder: (value: T) => Maybe<U>) => (self: Maybe<T>): Maybe<U> =>
+  isJust(self) ? binder(self.value) : nothing<U>();
+
+// --- Traversable ---
+
+// For a given predicate, this function finds out if there **exists** a value
+// inside this Maybe object that meets the predicate. In practice, this means
+// that for empty objects it returns false (as no predicate-meeting object
+// exists).
+export const exists = <T>(predicate: (value: T) => boolean) => (self: Maybe<T>): boolean =>
+  isJust(self) && predicate(self.value);
+
+// For a given predicate, this function finds out if **all** the values inside
+// this maybe object meet the predicate. In practice, this means that for empty
+// objects it returns true (as all 0 objects do meet the predicate).
+export const forAll = <T>(predicate: (value: T) => boolean) => (self: Maybe<T>): boolean =>
```

Comment:
Should this be `forall`. As every other API uses that casing and this seems like it'd get confusing.

Reply:
There's `Arr.forall`, but I don't know of others. If camelCase is our standard for function names, should we change `Arr.forall` to `forAll` instead?

Reply:
That'd be a breaking change and we didn't want to do that currently.

Reply:
We can alias it and migrate later. 

Equally, we can use `forall` everywhere, then change to `forAll` everywhere later on if we so desire. I'll leave it up to your team to decide.

Reply:
Fixed